bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieClient.java

```diff
@@ -20,579 +20,60 @@
-            callback.masterKey = masterKey;
-            callback.allowFastFail = allowFastFail;
-            callback.writeFlags = writeFlags;
-            return callback;
-        }
-
-        @Override
-        public void operationComplete(final int rc,
-                                      PerChannelBookieClient pcbc) {
-            if (rc != BKException.Code.OK) {
-                bookieClient.completeAdd(rc, ledgerId, entryId, addr, cb, ctx);
-            } else {
-                pcbc.addEntry(ledgerId, masterKey, entryId,
-                              toSend, cb, ctx, options, allowFastFail, writeFlags);
-            }
-
-            toSend.release();
-            recycle();
-        }
-
-        private ChannelReadyForAddEntryCallback(
-                Handle<ChannelReadyForAddEntryCallback> recyclerHandle) {
-            this.recyclerHandle = recyclerHandle;
-        }
-
-        private static final Recycler<ChannelReadyForAddEntryCallback> RECYCLER =
-            new Recycler<ChannelReadyForAddEntryCallback>() {
-                    protected ChannelReadyForAddEntryCallback newObject(
-                            Recycler.Handle<ChannelReadyForAddEntryCallback> recyclerHandle) {
-                        return new ChannelReadyForAddEntryCallback(recyclerHandle);
-                    }
-                };
-
-        public void recycle() {
-            bookieClient = null;
-            toSend = null;
-            ledgerId = -1;
-            entryId = -1;
-            addr = null;
-            ctx = null;
-            cb = null;
-            options = -1;
-            masterKey = null;
-            allowFastFail = false;
-            writeFlags = null;
-            recyclerHandle.recycle(this);
-        }
-    }
-
-    public void readLac(final BookieSocketAddress addr, final long ledgerId, final ReadLacCallback cb,
-            final Object ctx) {
-        final PerChannelBookieClientPool client = lookupClient(addr);
-        if (client == null) {
-            cb.readLacComplete(getRc(BKException.Code.BookieHandleNotAvailableException), ledgerId, null, null,
-                    ctx);
-            return;
-        }
-        client.obtain((rc, pcbc) -> {
-            if (rc != BKException.Code.OK) {
-                try {
-                    executor.executeOrdered(ledgerId, safeRun(() -> {
-                        cb.readLacComplete(rc, ledgerId, null, null, ctx);
-                    }));
-                } catch (RejectedExecutionException re) {
-                    cb.readLacComplete(getRc(BKException.Code.InterruptedException),
-                            ledgerId, null, null, ctx);
-                }
-            } else {
-                pcbc.readLac(ledgerId, cb, ctx);
-            }
-        }, ledgerId);
-    }
-
-    public void readEntry(BookieSocketAddress addr, long ledgerId, long entryId,
-                          ReadEntryCallback cb, Object ctx, int flags) {
+public interface BookieClient {
```

Comment:
since we are abstracting the interface, it is a good time to add comments for individual methods. it is an internal contract between components.